packages/ckeditor5-engine/src/view/observer/focusobserver.ts

```diff
@@ -48,35 +49,14 @@ export default class FocusObserver extends DomEventObserver<'focus' | 'blur'> {
 		this.useCapture = true;
 		const document = this.document;
 
-		document.on<ViewDocumentFocusEvent>( 'focus', () => {
-			this._isFocusChanging = true;
-
-			// Unfortunately native `selectionchange` event is fired asynchronously.
-			// We need to wait until `SelectionObserver` handle the event and then render. Otherwise rendering will
-			// overwrite new DOM selection with selection from the view.
-			// See https://github.com/ckeditor/ckeditor5-engine/issues/795 for more details.
-			// Long timeout is needed to solve #676 and https://github.com/ckeditor/ckeditor5-engine/issues/1157 issues.
-			//
-			// Using `view.change()` instead of `view.forceRender()` to prevent double rendering
-			// in a situation where `selectionchange` already caused selection change.
-			this._renderTimeoutId = setTimeout( () => {
-				this.flush();
-				view.change( () => {} );
-			}, 50 );
-		} );
-
-		document.on<ViewDocumentBlurEvent>( 'blur', ( evt, data ) => {
-			const selectedEditable = document.selection.editableElement;
-
-			if ( selectedEditable === null || selectedEditable === data.target ) {
-				document.isFocused = false;
-				this._isFocusChanging = false;
-
-				// Re-render the document to update view elements
-				// (changing document.isFocused already marked view as changed since last rendering).
-				view.change( () => {} );
+		document.on<ViewDocumentFocusEvent>( 'focus', () => this._handleFocus() );
+		document.on<ViewDocumentBlurEvent>( 'blur', ( evt, data ) => this._handleBlur( data ) );
+
+		document.on<ViewDocumentInputEvent>( 'beforeinput', () => {
```

Comment:
Comments needed. Why we need to handle focus on every beforeinput? It was confusing initially that this actually SHOULD happen only once. Because it relies on the assumption that beforeinput is fired only for focused editors. Or not? If not, what happens then?

Reply:
Q: What happens when beforeinput is fired for a blurred editable?

Reply:
What happens when `beforeinput` is fired for a blurred editable is described in this issue: https://github.com/ckeditor/ckeditor5/issues/14702.

Reply:
Added in ce023f1.